[Aditi Gawande] iP

aditig0305 opened the pull request:
### MAX
> “Your mind is for having ideas, not holding them.” – David Allen

Max frees your mind from the burden of remembering tasks. It's:
- text-based
- easy to learn
- FAST—**SUPER FAST** to use

All you need to do is:
1. download it from [here](https://github.com/aditig0305/ip).
2. clone the repository
3. run the code
4. add your tasks
5. let it manage your tasks for you 😉

And it is **_FREE_**!

Features:
- [ ] Managing tasks
- [ ] Adding deadlines
- [ ] Searching for tasks 

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:


```java
public static void main(String[] args) {
    Max max = new Max();
    max.runMax();
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

Commit:
Rename, Greet, Exit

Commit:
Level 1 Echo

Commit:
Level 2 - Add, List

Commit:
Level 3 - Mark as Done

Commit:
Level 4 - ToDos, Events, Deadlines

Commit:
Automated Text UI Testing and Level 5 - Handle Errors

Commit:
Level 6 - Delete

Commit:
Level 7 - Save

Commit:
Reorganised Max

Commit:
Merge branch 'branch-Level-7'

Added save feature and reorganised Max.java

Commit:
Almost done with Level 8. Deadline task accepts both a date that it can parse and a string.

Commit:
Deadline handling: Support date formats

Teach the chatbot to understand and handle multiple date formats for
deadline tasks. This enhancement allows the chatbot to parse and
interpret dates provided in different formats, making it more
flexible and user-friendly.

The chatbot now supports:
* Input format: d/M/yyyy HHmm (e.g., '2/12/2019 1800')
* Output format: MMM dd yyyy HH:mm (e.g., 'Dec 02 2019 18:00')

The date is stored as a LocalDateTime object in the task. The update
ensures correct parsing and formatting of date strings for deadlines
and improves the chatbot's ability to manage date-related tasks.

This change includes:
* Parsing date strings in multiple formats.
* Storing and formatting dates in a consistent manner.

Commit:
Merge branch 'branch-Level-8'

Commit:
Refactor code to enhance OOP principles: Modularize classes, improve encapsulation, and separate concerns

- Split the `Max` class into multiple classes (`Ui`, `Storage`, `Parser`, `TaskList`) to follow the Single Responsibility Principle.
- Improved encapsulation by making class attributes private and providing getter/setter methods where necessary.
- Moved task-related methods to the `TaskList` class to centralize task management.
- Updated the `Storage` class to handle file I/O operations independently.
- Simplified `Parser` class logic for handling user inputs and parsing dates.
- Improved overall code readability and maintainability.

Commit:
Add packages for better organisation and OOP improvements

The project has a flat directory structure under src/main/java which makes it challenging to manage and scale.

Organising the code into packages improves modularity and readability. It allows better separation of concerns, making the code easier to maintain and navigate.

Created packages for different components of the project:

package task; for classes related to task management (Task, Event, Deadline, Todo).
package exception; for custom exceptions (MaxException).
package main; for the main application entry point and other logic (Max, Parser, Storage, Ui).
Updated import statements to reflect new package structure.

Refactored classes to use the new package names.

Ensured that all references between classes are updated to match the new package structure.

Using packages aligns with best practices in object-oriented programming by grouping related classes together. It simplifies navigation, reduces classpath conflicts, and makes it easier to scale the application.

Commit:
Merge branch 'add-gradle-support'

Commit:
Integrate Gradle for automated project builds

The project previously relied on manual build processes. To streamline
and automate these tasks, Gradle has been integrated into the project.
This change allows for consistent builds, easy dependency management,
and simplified project setup.

- Merged the `add-gradle-support` branch to introduce Gradle files.
- Set up Gradle wrapper and build script.
- Verified that the project can be built and run using Gradle.

This integration will improve development efficiency by automating
build tasks and ensuring consistency across different environments.

Commit:
Add JUnit tests for Parser and Todo

The Parser and Todo classes lacked comprehensive unit tests,
increasing the risk of undetected bugs, especially in date parsing
and task validation.

To address this, the following changes were made:
- Added JUnit tests for the parseDate method in Parser:
- Added JUnit tests for the Todo class

These changes improve the robustness of the Parser and Todo
classes, ensuring they handle various inputs correctly.

Commit:
Package app as executable JAR file

The app is currently not packaged in a way that allows for easy distribution
and execution by end-users.

To improve usability, package the app as an executable JAR file. This allows users
to run the application simply by copying the JAR file into an empty folder,
opening a command window in that folder, and executing the command:
java -jar "Duke.jar"

This ensures that the app can be distributed and executed without
requiring additional setup, making it more user-friendly and accessible.

Let's configure the build process to generate the JAR file as Duke.jar
and place it in the build/libs directory for consistency and ease of access.

Commit:
Tweak code to comply with coding standard

The existing codebase does not fully adhere to the given coding standard, which could lead to inconsistencies and maintenance challenges.

To improve code quality and maintainability, tweak the existing code to comply with the specified coding standard. From this point onward, ensure that any new code added is compliant with this standard.

By enforcing a consistent coding standard across the project, we reduce the risk of bugs and make the codebase easier to read and maintain. Let's apply these changes systematically to ensure uniformity throughout the project.

Commit:
Add search functionality to find tasks by keyword

The application currently lacks a feature to search for tasks by a keyword, making it difficult for users to quickly locate specific tasks in their list.

To enhance usability, introduce a find command that allows users to search for tasks containing a specified keyword. This feature will display all matching tasks in a user-friendly format, making task management more efficient.

Implementing this functionality improves the overall user experience by allowing users to easily filter and find relevant tasks based on their search criteria.

Commit:
Add JavaDoc comments to all classes

Add JavaDoc comments to all classes in the project for improved documentation and clarity.

- Added descriptions and parameter explanations to the Max class.
- Documented methods and constructors in the Parser class.
- Included JavaDoc for all methods and fields in the Storage class.
- Added comments to the Event, Todo, and Task classes explaining their functionality.
- Documented methods and parameters in the TaskList class.

This update helps to provide a better understanding of the codebase and improves maintainability.

Commit:
Merge branch 'branch-A-CodingStandard'

Commit:
Merge branch 'branch-A-JavaDoc'

# Conflicts:
#	src/main/java/max/main/Storage.java

Commit:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/max/main/Parser.java
#	src/main/java/max/main/Ui.java

Commit:
Add JavaDoc comments for printList and handleFind methods

Add JavaDoc comments to the following methods:
* printList: Documented its functionality to print tasks with or without filters based on the isFilter parameter.
* handleFind: Explained its role in filtering tasks by keyword, including parameter details and exception handling.

This improves code readability and provides clear documentation for future reference.

Commit:
Integrate JavaFX for GUI support

Commit:
Integrate JavaFX for GUI support

Add JavaFX dependency and set up the initial GUI framework.

JavaFX integration introduces a graphical user interface to replace the console-based interaction.

Here's what was done:

Added JavaFX libraries to the project.
	Created Main class to launch the JavaFX application.
	Implemented MainWindow class with FXML-based UI components.
	Updated Max class to interact with the new JavaFX-based UI.
	Added MainWindow.fxml to define the layout of the main window.
	Configured MainWindow controller to handle user inputs and display responses.
This setup allows for a more interactive user experience and prepares the project for further GUI enhancements.

Commit:
Merge branch 'branch-Level-10'

jloh02 commented on src/main/java/max/main/DialogBox.java:
```diff
@@ -0,0 +1,60 @@
+package max.main;
+
+import java.io.IOException;
+import java.util.Collections;
+
+import javafx.collections.FXCollections;
+import javafx.collections.ObservableList;
+import javafx.fxml.FXML;
+import javafx.fxml.FXMLLoader;
+import javafx.geometry.Pos;
+import javafx.scene.Node;
+import javafx.scene.control.Label;
+import javafx.scene.image.Image;
+import javafx.scene.image.ImageView;
+import javafx.scene.layout.HBox;
```
Great job keeping your import order consistent and in alphabetical within each section

jloh02 commented on src/main/java/max/main/Main.java:
```diff
@@ -0,0 +1,42 @@
+
+    private Max max = new Max();
+
+    /**
+     * The main entry point for the JavaFX application. It initializes the primary stage,
+     * loads the FXML layout for the main window, and sets the scene for the stage.
+     * It also injects the Max instance into the MainWindow controller to enable interaction
+     * between the GUI and the application logic.
+     *
+     * @param stage The primary stage for this application, onto which the application scene will be set.
+     */
```
Nice descriptive JavaDoc comment with appropriate parameter description

jloh02 commented on src/main/java/max/main/Parser.java:
```diff
@@ -0,0 +1,205 @@
+package max.main;
+
+import max.exception.MaxException;
+import max.task.*;
```
Any particular reason a wildcard import was used here?

jloh02 commented on src/main/java/max/main/Storage.java:
```diff
@@ -0,0 +1,106 @@
+import max.task.Task;
+import max.task.Todo;
+
+import java.io.*;
```
Similarly, would be great to avoid wildcard imports and list imports explicitly

jloh02 commented on src/main/java/max/main/Max.java:
```diff
@@ -0,0 +1,59 @@
+    private Ui ui;
+
+    private boolean isExit = false;
+    private static final String FILE_PATH = "./data/max.txt";
```
Great job keeping modifiers order in accordance to the coding standard

tsulemornorfi commented on src/main/java/max/main/Storage.java:
```diff
@@ -0,0 +1,106 @@
+            while ((line = reader.readLine()) != null) {
+                String[] parts = line.split(" \\| ");
+                Task task;
+                switch (parts[0]) {
```
cases should not be indented

currynia reviewed (commented):
LGTM

Commit:
Refactor imports and fix code indentation

Replaced wildcard imports with explicit imports to improve code clarity and maintainability.
Corrected inconsistent indentations to adhere to coding standards and enhance readability.
These changes contribute to a cleaner and more maintainable codebase.

tsulemornorfi commented on src/main/java/max/task/Deadline.java:
```diff
@@ -0,0 +1,66 @@
+package max.task;
+
+import max.task.Task;
```
this import is redundant

tsulemornorfi commented on src/main/java/max/task/Event.java:
```diff
@@ -0,0 +1,53 @@
+package max.task;
+
+import max.exception.MaxException;
+import max.task.Task;
```
This import is redundant

tsulemornorfi commented on src/main/java/max/main/Max.java:
```diff
@@ -0,0 +1,59 @@
+import max.exception.MaxException;
+import max.task.TaskList;
+
+import java.util.Scanner;
```
java.util.Scanner should before those in line 3 and 4.

tsulemornorfi commented on src/main/java/max/main/Parser.java:
```diff
@@ -0,0 +1,205 @@
+import java.time.format.DateTimeFormatter;
+import java.time.format.DateTimeParseException;
+import java.util.ArrayList;
+import java.util.Scanner;
```
java.util.* should come before those in 3 and 4.

currynia reviewed (commented):
LGTM mostly very minor issues

Commit:
Add assertions to enforce input validation and task constraints across classes

- **Parser class:**
  - Asserted that the `taskList`, `ui`, and `storage` objects are not null in the constructor to
	ensure these components are initialised before use.
  - Added assertions to validate non-null and non-empty strings in methods like `handleDeadline`,
	`handleTodo`, and `handleEvent` to prevent invalid task descriptions.
  - Asserted that the task index used in methods like `mark`, `unmark`, and `delete` is within
	valid bounds to avoid IndexOutOfBoundsException.
- **Ui class:**
  - Asserted that the task list size is non-negative in methods like `printTaskTypeAdded` and `printDeleteTask` to prevent invalid states.
  - Ensured tasks are not null before printing them in methods like `printMarkDone` and `printMarkNotDone`.
  - Added checks to ensure that the task list passed to the `list` method is not null, preventing null pointer dereferencing.

These assertions improve robustness by validating key preconditions, catching errors early, and preventing common runtime exceptions such as null pointers or invalid indices.

Commit:
Refactor long methods and remove unused imports

Current Situation:
- Several methods in the codebase, particularly in the Parser and task handling, were long and doing multiple things at once.
- This made the methods difficult to read, understand, and maintain.
- Some imports were present in the codebase but were no longer being used, adding unnecessary clutter.

Why It Needs to Change:
- Long methods often violate the Single Responsibility Principle (SRP) and can make the code harder to test.
- Refactoring them into smaller, more modular methods improves readability, maintainability, and allows for easier debugging and testing.
- Unused imports increase clutter and may lead to confusion or performance issues, making the codebase less clean.

What is Being Done:
- Refactored larger methods such as `parseText`, `handleMark`, `handleUnmark`, and others into smaller helper methods like `handleList`, `handleDelete`, etc.
- Extracted distinct logical units from long methods into smaller methods with clear, focused responsibilities.
- Updated corresponding Javadocs to document these changes.
- Removed all unused imports to declutter the code and avoid any unnecessary dependencies.

Why It Is Done This Way:
- Breaking down the logic into smaller, well-named methods makes it easier to follow and modify the code without introducing errors.
- This refactoring adheres to the principle of writing cleaner, more modular code.
- Removing unused imports enhances code cleanliness and efficiency, contributing to better overall code hygiene.

Any Other Relevant Info:
- No changes in functionality; the refactoring is purely for improving code structure and readability.

aditig0305 committed:
Merge pull request #1 from aditig0305/branch-A-Assertions

Add assertions to enforce input validation and task constraints acros…

Commit:
Merge branch 'master' into branch-A-CodeQuality

aditig0305 committed:
Merge pull request #2 from aditig0305/branch-A-CodeQuality

Refactor long methods and remove unused imports

Commit:
Add help page with command descriptions

- Implemented a help page for Max to assist users with available commands.
- Updated the print formatting to clearly list commands and their descriptions.

This update enhances user experience by providing clear guidance on using Max effectively.

Commit:
Merge branch 'branch-C-Help'

Commit:
Add Tagging Functionality to Max

Features Added:
- Implemented the ability to add tags to tasks with the tag command.
- Implemented the ability to remove tags from tasks with the untag command.
- Added search functionality for tasks by tag with the searchtag command.
- Updated the printHelp method to include instructions for new tag-related commands.

Tagging Implementation:
- handleTag(String text): Parses command input to extract task index and tag, adds the tag to the specified task, and provides feedback to the user.
- handleUntag(String text): Parses command input to extract task index and tag, removes the tag from the specified task, and provides feedback to the user.
- handleSearchtag(String text): Parses command input to extract the tag, searches for tasks containing the specified tag, and displays the results.

Help Page Update:
- Added descriptions for new tag-related commands (tag, untag, searchtag) to the printHelp method.

Error Handling:
- handleTag and handleUntag methods now throw MaxException if the command input is incomplete.
- handleSearchtag checks for valid tag input and handles errors appropriately.

Code Changes:
- Modified Task class to include methods for adding and removing tags.
- Updated Storage class to handle tag in file format.

Commit:
Merge branch 'branch-C-Tagging'

Commit:
Add padding and anchoring to improve GUI layout

Current situation: The GUI elements lacked padding, leading to text and images being tightly packed, which reduced readability and visual appeal. Anchoring was not used effectively, causing inconsistent alignment across different window sizes.

Why it needs to change: Padding and anchoring are essential for creating a clean, well-organized user interface. This enhances the user experience by improving readability, spacing, and overall layout consistency.

What was done:
Added padding to Label and ImageView components, ensuring that text and images are spaced out properly for better readability.
Applied anchoring to align elements more consistently within the dialog boxes, especially when resizing the window.

Why it was done this way: The padding and anchoring ensure a clean and organized visual structure, enhancing both usability and aesthetic appeal. These changes also provide better layout control, making the GUI adaptable to different screen sizes.

Commit:
Merge branch 'tweak-ui'

Commit:
Update README with Comprehensive User Guide

- Added a detailed user guide to the README file to help users effectively navigate and utilize the Max task manager application.
- Included sections for:
  - **Adding Tasks:** Instructions for adding to-do items, deadlines, and events.
  - **Managing Tasks:** Commands to mark tasks as done or not done, delete tasks, list tasks, and find tasks by keyword.
  - **Tags:** How to add, search, and remove tags from tasks.
  - **Help:** Command to display the user guide.

- Commands are formatted using backticks for better readability.
- Examples for each command provided to demonstrate usage.
- Structured the guide into clear sections with a Table of Contents for easy navigation.

Commit:
Remove some comments from README

Added a detailed user guide to the README file to help users effectively navigate
and utilize the Max task manager application.

Commit:
Refactor parseText and printHelp methods for simplicity and clarity

Current situation:
The parseText and printHelp methods were long, containing repetitive code for handling
various commands and printing help text. This caused unnecessary duplication and made the
 code harder to maintain or extend.

Why it needs to change:
Lengthy methods can reduce readability and increase maintenance costs when adding new commands.
Additionally, repetitive code violates the DRY (Don't Repeat Yourself) principle.

What is being done:
The parseText method is refactored using helper methods (isSimpleCommand, isIndexCommand,
handleSimpleCommand, etc.) to handle different command types more modularly.
The printHelp method now uses an array of command-descriptions pairs to remove redundancy when
printing the help text.
SLAP (Single Level of Abstraction Principle) is applied to better structure both methods, ensuring each
method handles a single level of responsibility.

Why it is done that way:
This modular approach simplifies command parsing in parseText and reduces duplication in printHelp.
By using helper methods and arrays, it becomes easier to add, update, or remove commands without
affecting the structure of the main methods.

Other relevant info:
Additional methods (isByeCommand, handleSimpleCommand, etc.) improve the maintainability of the code.
Future command extensions will be easier to implement with these abstractions.

Commit:
Add TaskList junit tests

Commit:
Debug storage error for events

Commit:
Add exception handling to task list

Commit:
Handle commands when no tasks

Commit:
Add tweaks to exception message

aditig0305 committed:
Improve read me
